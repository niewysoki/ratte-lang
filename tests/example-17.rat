fn HigherOrder(f: fn (Int, Int) -> Int, x: Int, y: Int) -> Int {
  return f(x, y);
}

fn Clojure(x0: Int) -> fn (Int) -> Int {
  let mut y: Int = x0;
  return | x: Int | -> Int  { y = y + x; return y; };
}

fn Add(x: Int, y: Int) -> Int {
  return x + y;
}

fn Lambdas() -> Void {
  let x: Int = 14;
  let lambda1: fn(Int) -> Int = | x: Int | -> Int  { return x + 1; };

  Println(lambda1(x));

  let lambda2: fn(Int) -> fn(Int) -> Int = | x: Int | -> fn (Int) -> Int  {
    return | y: Int | -> Int { return Add(x, y); };
  };


  let fun1: fn(Int) -> Int = lambda2(1);
  let fun2: fn (Int) -> Int = lambda2(3);
  Println(fun1(16));
  Println(fun2(16));

  return;
}


fn Main() -> Void {
  Println(HigherOrder(Add, 4, 5));
  let c: fn(Int) -> Int = Clojure(10);
  Println(ShowInt(c(1)));
  Println(ShowInt(c(2)));
  Lambdas();
  return;
}