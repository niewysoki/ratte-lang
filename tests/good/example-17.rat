fn Apply (f: fn (mut Int) -> Int, mut x: Int) -> Int {
  return f(x);
}

fn Clojure (init: Int) -> fn (Int) -> Int {
  let mut x = init;
  return | y: Int | -> Int {
    x = x + y;
    return x;
  };
}

fn Inc (x: Int) -> Int {
  return x + 1;
}

fn IncMut (mut x: Int) -> Int {
  x++;
  return x;
}

fn Main() -> Void {
  let mut x = 0;
  Println(ShowInt(Apply(Inc, x)));
  Println(ShowInt(Apply(IncMut, x)));
  Println(ShowInt(Apply(Inc, x)));
  let counter = Clojure(10);
  Println(ShowInt(counter(1)));
  Println(ShowInt(counter(3)));
  Println(ShowInt(counter(5)));

  // value args can be promoted to reference args
  let c1: fn (mut Int) -> fn (mut Int) -> Int = Clojure;
  let c2: fn (mut Int) -> fn (Int) -> Int = Clojure;
  let c3: fn (Int) -> fn (mut Int) -> Int = Clojure;
  let c4: fn (Int) -> fn (Int) -> Int = Clojure;

  fn ClojureMut (mut x : Int) -> fn (mut Int) -> Int {
    return Clojure(x);
  };

  // reference args cannot be promoted to value args
  let cMut: fn (mut Int) -> fn (mut Int) -> Int = Clojure;

  return;
}