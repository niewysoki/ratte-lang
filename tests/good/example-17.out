1
1
2
11
14
19
